File: day_1/day_1.py
from collections import Counter

def compute_total_distance_and_similarity_score(file_path):
    left_list = []
    right_list = []
    
    # Read and process the file line by line
    with open(file_path, 'r', encoding='utf-8-sig') as file:
        for line in file:
            if line.strip():  # Skip empty lines
                left, right = map(int, line.split())
                left_list.append(left)
                right_list.append(right)
    
    # Sort both lists for total distance calculation
    left_list_sorted = sorted(left_list)
    right_list_sorted = sorted(right_list)
    total_distance = sum(abs(l - r) for l, r in zip(left_list_sorted, right_list_sorted))
    
    # Count occurrences in the right list for similarity score
    right_count = Counter(right_list)
    similarity_score = sum(num * right_count[num] for num in left_list)
    
    return total_distance, similarity_score


file_path = './input.txt'
total_distance, similarity_score = compute_total_distance_and_similarity_score(file_path)

# Print the results
print("Total distance:", total_distance)
print("Similarity score:", similarity_score)

File: day_1/day_1.js
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
function parseInput(filePath) {
    var data = fs.readFileSync(filePath, 'utf-8').trim();
    var leftList = [];
    var rightList = [];
    for (var _i = 0, _a = data.split('\n'); _i < _a.length; _i++) {
        var line = _a[_i];
        var _b = line.split(/\s+/).map(Number), left = _b[0], right = _b[1];
        leftList.push(left);
        rightList.push(right);
    }
    return { leftList: leftList, rightList: rightList };
}
function calculateTotalDistance(leftList, rightList) {
    var sortedLeft = __spreadArray([], leftList, true).sort(function (a, b) { return a - b; });
    var sortedRight = __spreadArray([], rightList, true).sort(function (a, b) { return a - b; });
    var totalDistance = 0;
    for (var i = 0; i < sortedLeft.length; i++) {
        totalDistance += Math.abs(sortedLeft[i] - sortedRight[i]);
    }
    return totalDistance;
}
function calculateSimilarityScore(leftList, rightList) {
    var rightFrequency = {};
    for (var _i = 0, rightList_1 = rightList; _i < rightList_1.length; _i++) {
        var num = rightList_1[_i];
        rightFrequency[num] = (rightFrequency[num] || 0) + 1;
    }
    var similarityScore = 0;
    for (var _a = 0, leftList_1 = leftList; _a < leftList_1.length; _a++) {
        var num = leftList_1[_a];
        if (rightFrequency[num]) {
            similarityScore += num * rightFrequency[num];
        }
    }
    return similarityScore;
}
function main() {
    var filePath = './input.txt';
    var _a = parseInput(filePath), leftList = _a.leftList, rightList = _a.rightList;
    var totalDistance = calculateTotalDistance(leftList, rightList);
    var similarityScore = calculateSimilarityScore(leftList, rightList);
    console.log('Total Distance :', totalDistance);
    console.log('Similarity Score :', similarityScore);
}
main();
